apps/launcher/datafilespage.cpp

```diff
@@ -308,3 +326,23 @@ bool Launcher::DataFilesPage::showDeleteMessageBox (const QString &text)
 
     return (msgBox.clickedButton() == deleteButton);
 }
+
+void Launcher::DataFilesPage::slotAddonDataChanged()
+{
+    QStringList selectedFiles = selectedFilePaths();
+    if (previousSelectedFiles != selectedFiles) {
+        previousSelectedFiles = selectedFiles;
+        // Loading cells for core Morrowind + Expansions takes about 0.2 seconds, which is enough to cause a
+        // barely perceptible UI lag. Splitting into its own thread to alleviate that.
+        std::thread loadCellsThread(&DataFilesPage::reloadCells, this, selectedFiles);
+        loadCellsThread.join();
```

Comment:
I fail to see how this is any different than simply calling `reloadCells(selectedFiles);`. You spawn a thread and wait until it finishes while still blocking the UI...

Reply:
The intention was definitely for `reloadCells(selectedFiles()` to be called in a separate thread while not blocking UI. Is there a better way that I could do this so that UI is not blocked?

Reply:
You can remove the `join` call and the main thread won't be blocked.

However, I see a potential problem with `reloadCells` running in the background. If I understand correctly, if the user changes mods multiple times in quick succession, multiple threads will run simultanenously and there is no telling in which order the `signalSelectedFilesChanged` signal will be emitted. Therefore the list of cells might end up in an inconsistent state.

Reply:
@MiroslavR I tried removing `join`, but got `SIGABRT (Signal 6)` and the app crashed. In looking through C++ documentation, I'm guessing that instead I want to replace `join` with [`detatch`](http://www.cplusplus.com/reference/thread/thread/detach/). Is that correct?

Good point about the race condition with regards to results. I couldn't trigger one on my own, but with a large number of mods, it's certainly possible if a big one (e.g. Tamriel Rebuilt) is enabled or disabled in quick succession. Is implementing a queue of some sort the best way to handle this, or is there a cool C++ feature that I can use that would be better?

Reply:
> @MiroslavR I tried removing join, but got SIGABRT (Signal 6) and the app crashed. In looking through C++ documentation, I'm guessing that instead I want to replace join with detatch. Is that correct?

I edited my post.

> Is implementing a queue of some sort the best way to handle this, or is there a cool C++ feature that I can use that would be better?

Honestly I'm not sure what the best approach would be. I suppose you could also disable content selection for the duration of the cell loading thread. Or get rid of threading altogether if the delay is barely noticeable. Maybe @zinnschlag or @scrawl could chime in?

Reply:
I replaced `join` with `detach` so that the UI will not be blocked.